Add sco connection management

lylezhu2012 opened the pull request:
Currently, SCO connections and disconnections are agnostic to upper-layer.

Add two functions, bt_sco_connected and bt_sco_disconnected, to notify the SCO connect changes.

For Central side, pass "struct bt_sco_chan" object when calling bt_conn_create_sco. it uses to manage the SCO channel for upper-
layer.

For Peripheral side, two functions bt_sco_server_register and bt_sco_server_unregister are added to monitor SCO connection request for upper-layer. The upper-layer could accept or reject SCO connect When the connection request received. If the connect is accepted, the "struct bt_sco_chan" object could be passed by "sco_server->accept".

Call bt_sco_server_register to register SCO server. Manage the SCO connection request. Get SCO connect/disconnect status via SCO channel ops.

Notify the uppper layer of the SCO connection status changes through bt_hfp_hf_cb::sco_connected and bt_hfp_hf_cb::sco_disconnected.

lylezhu2012 committed:
Bluetooth: Host: Add SCO connection management

Currently, SCO connections and disconnections
are agnostic to upper-layer.

Add two functions, bt_sco_connected and
bt_sco_disconnected, to notify the SCO
connect changes.

For Central side, pass "struct bt_sco_chan"
object when calling bt_conn_create_sco. it
uses to manage the SCO channel for upper-
layer.

For Peripheral side, two functions
bt_sco_server_register and
bt_sco_server_unregister are added to monitor
SCO connection request for upper-layer. The
upper-layer could accept or reject SCO connect
When the connection request received. If the
connect is accepted, the "struct bt_sco_chan"
object could be passed by "sco_server->accept".

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Host: Add SCO Connect notify for HFP

Call bt_sco_server_register to register SCO
server. Manage the SCO connection request.
Get SCO connect/disconnect status via SCO
channel ops.

Notify the uppper layer of the SCO connection
status changes through bt_hfp_hf_cb::
sco_connected and bt_hfp_hf_cb::
sco_disconnected.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jori-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1774,6 +1774,66 @@ struct bt_br_conn_param {
+ *  dealing with setting channel to proper state depending on operational
+ *  context.
+ */
+enum bt_sco_state {
```
If it's only used by internal APIs, could it be moved to `conn_internal.h`?
If it's designed to be used from an upper layer (possibly not in the zephyr tree), this line needs to be reworded.

lylezhu2012 replied:
Due to the new parameter `struct bt_sco_chan *chan` is added to function `bt_conn_create_sco`, this enum needs to be defined in same file `conn.h`.
If function `bt_conn_create_sco` could be moved to `conn_internal.h` or `sco_internal.h`, this new added code block could be moved to corresponding file.

jori-nordic replied:
Okay. After reading your explanation above, I would then prefer the SCO API be internal for now.
Then `conn.h` doesn't need to be modified.

lylezhu2012 replied:
All of changes are moved to `sco_internal.h`

jori-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1096,7 +1097,10 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 	case BT_CONN_CONNECTED:
 		if (conn->type == BT_CONN_TYPE_SCO) {
-			/* TODO: Notify sco connected */
+			/* Notify sco connected */
```
nit: You probably don't need the comment anymore.

lylezhu2012 replied:
Removed

jori-nordic commented on subsys/bluetooth/host/sco_internal.h:
```diff
@@ -0,0 +1,96 @@
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_sco_server_register(struct bt_sco_server *server);
```
Don't address this in this PR, it's just something I think that should be done in the future.

To me that sounds like an API that should be public. I don't know enough about Classic, are users allowed to create SCO connections?
In the same vein, I see that in order to use HFP, etc.. you have to include private headers, ie the tests that test RFCOMM and other classic profiles need to include the corresponding private header for the profile API.

I believe moving those to a `include/bluetooth/classic/[profilename].h` would be a good thing for usability.
Unless they are not supposed to be configured from the app, tell me if I'm wrong @jhedberg .

jhedberg replied:
> I believe moving those to a `include/bluetooth/classic/[profilename].h` would be a good thing for usability. Unless they are not supposed to be configured from the app, tell me if I'm wrong @jhedberg.

Yes, I think we're overdue with having dedicated directories for BR/EDR implementation and header files. It'd also make the references in `MAINTAINERS.yml` simpler. So I'd propose to first move all existing BR/EDR-specific code to `subsys/bluetooth/host/classic/` and `include/bluetooth/classic/` before adding any new files or features.

lylezhu2012 replied:
From my view, SCO and eSCO need to work together with HFP unit/HFP AG. So I think the source code of SCO/eSCO could be moved internally without exposing the interface. The SCO/eSCO connection will be opened or closed internally by the HFP unit/AG following the HFP sequence in a timely manner.

I will do it in this PR.

> Yes, I think we're overdue with having dedicated directories for BR/EDR implementation and header files. It'd also make the references in `MAINTAINERS.yml` simpler. So I'd propose to first move all existing BR/EDR-specific code to `subsys/bluetooth/host/classic/` and `include/bluetooth/classic/` before adding any new files or features.
Yes. What can I do to help make this change happen?

jori-nordic replied:
You can just make another PR moving the files (and updating the paths in include, cmake and Maintainers)

jori-nordic reviewed (changes requested):
Blocking b/c of the `conn.h` wording. All other comments are not blocking.

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1774,6 +1774,66 @@ struct bt_br_conn_param {
+
+	enum bt_sco_state		state;
+
+#if defined(CONFIG_BT_SMP) || defined(__DOXYGEN__)
+	/** @brief The required security level of the channel
+	 *
+	 * This value can be set as the central before calling bt_conn_create_sco().
+	 * The value is overwritten to @ref bt_iso_server::sec_level for the
+	 * peripheral once a channel has been accepted.
+	 *
+	 * Only available when @kconfig{CONFIG_BT_SMP} is enabled.
+	 */
+	bt_security_t			required_sec_level;
+#endif /* CONFIG_BT_SMP */
```
Do we really need to track this independently from the parent ACL connection?

lylezhu2012 replied:
Currently, security related checks are not necessary. I removed it.

jhedberg commented on subsys/bluetooth/host/classic/sco.c:
```diff
@@ -0,0 +1,409 @@
+		return -EINVAL;
+	}
+
+#if defined(CONFIG_BT_SMP)
+	if (server->sec_level > BT_SECURITY_L3) {
+		return -EINVAL;
+	}
+#endif /* CONFIG_BT_SMP */
```
This part doesn't make sense to me. BR/EDR will in practice always have security support available, and it's not dependent on whether SMP is available (which is generally only used for LE).

lylezhu2012 replied:
L2CAP fixed channel of BR/EDR Security Manager is also implemented in `smp.c`. The file `smp.c` depends on `BT_SMP`. In current Kconfig, `BT_SMP` is selected by `BT_CLASSIC`. 
If the preprocessor directive `#if defined(CONFIG_BT_SMP)` ... `#endif /* CONFIG_BT_SMP */` is removed, it has no impact on compilation. But from the dependency perspective, the preprocessor directive `#if defined(CONFIG_BT_SMP)` ... `#endif /* CONFIG_BT_SMP */` is required.

jhedberg replied:
The BR/EDR SMP channel exists only for cross-transport key derivation. The actual BR/EDR pairing happens using BR/EDR-only methods (i.e. through Secure Simple Pairing or legacy BR/EDR pairing). The ability to establish a secure connection for BR/EDR, or the existence of a minimum security level requirement do not depend on SMP, so it looks illogical to have such dependencies in the code.

Whatever the underlying cause is (since I didn't completely understand it) for you wanting to have these #ifdefs should be fixed instead.

lylezhu2012 replied:
I think the preprocessor directive can be removed for BR/EDR. I will update it.

jhedberg commented on subsys/bluetooth/host/classic/sco.c:
```diff
@@ -0,0 +1,409 @@
+		return BT_HCI_ERR_SUCCESS;
+	}
+
+#if defined(CONFIG_BT_SMP)
+	if (conn->sec_level >= sco_server->sec_level) {
+		return BT_HCI_ERR_SUCCESS;
+	}
+
+	return BT_HCI_ERR_INSUFFICIENT_SECURITY;
+#else
```
Same here. BR/EDR security should not depend on SMP in any way (well, except for cross-transport key derivation, but that's a different story)

lylezhu2012 committed:
Bluetooth: Host: Remove SCO security check

Remove SCO security management code block.
Currently, security related checks are not
necessary. It can be added if needed.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>
